### DISCUSSION TITLE
Support for playlists

### DISCUSSION THREAD

**Discussion by beweavek** [meltaxa/jukebox-portal#7] [Ideas]
This is a great idea and I can't wait to start this project with my kids!

We don't have a huge collection of figures, so I was wondering if it could be modified to trigger a playlist of songs instead of a single track?  Either offline mp3 or an existing Spotify playlist?

Thanks for the great work!

**Comment by meltaxa** [comment]
Hi. 

Playlists are a great idea. Thankfully, Spotify playlists are already supported by Musicfig. See the [tags.yml-sample](https://github.com/meltaxa/musicfig/blob/master/tags.yml-sample#L18) file which list some possible combinations of how tags and songs can be associated. 

Simply copy the Spotify URI for the playlist and paste that in the tags files (add the space after spotify:). For example:
```
  ea508007:
        name: Iron DJ triathlon playlist
        spotify: playlist:2srSxWasD4ZZh4B7ckyihH
```

I'll look into how to support MP3 playlists in a future release.

In the meantime, have fun building Musicfig with the kids. My 7 years old loves it. It is a great learning tool too, with some of his tags playing audiobooks, educational tracks and songs.

**Reply by miky2184** [reply]
Hi...i have forked the project and I'm trying to support playing randomly a mp3 folder.

**Reply by meltaxa** [reply]
Great. Consider defining the songs or folders for the playlist in the tags.yml file, for example:
```
ea508007:
    name: my mp3 playlist
    mp3:
      - song1.mp3
      - song2.mp3
      - song3.mp3
      - path/to/album/folder
    shuffle: on
```
Submit a PR once you've got something.

**Comment by meltaxa** [comment]
The dev branch has an update that controls the lightshow based on the Spotify playing status. That is, if Spotify is playing a track, the lightshow will be on. If Spotify is paused or stopped, the lights are switched off. This will be merged to master soon based on further tests and feedback.

**Comment by miky2184** [comment]
Hi…
 I’ve done…the code is not written very well…is in alpha status…but working :)

I have defined a new “command” playlist ..so the tags.yml is:

Xxxxxxxxx:
	name: my mp3 playlist
	playlist: /path/to/album/folder

I’ve added a new queue —> playlist_queue , a new state PLAYLIST and a new Command PLAYLIST
And create a new fn _play_playlist


The code is the branch develop, if you want check my code before the pr or tell me another branch where I can do a pr.

Thank you

Bye Bye

**Reply by meltaxa** [reply]
In line 143, use "os.path.dirname(os.path.abspath(__file__)) + '/../music/'" ... in your path instead of hard-coding the music path (/home/pi/Music/). 

Or perhaps create a config item in tags.yml for the mp3 base dir, for example:
```
mp3_dir=/home/pi/Music
```
Then use mp3_dir  in the appropriate places in the code. If it is not defined, then it could default to the os.path.dirname(os.path.abspath(__file__)) + '/../music/'" path.

I'll endeavour to test your branch next.

**Comment by miky2184** [comment]
ok..i fixed the path...
i put in tags.yml

# Base path of mp3 folder
mp3_dir: /path/to/base/mp3/folder
i made a pr.

let me know.

bye bye
*Michele Micunco*


Il giorno mer 24 mar 2021 alle ore 00:54 Dennis Mellican <
***@***.***> ha scritto:

**Comment by r04drunn3r79** [comment]
Great app. 
Running the dev version, the lightshow keeps playing after every track on the playlist.

I do see a lot of errors in the musicfig.log file regarding the Spotify API.
Did I configure or installed something the wrong way?

Partial output 'musicfig.log':

```
[2021-03-27 17:37:26,812] ERROR: Spotify could not find any track playing: Error in 
https ://api.spotify.com/v1/me/player/currently-playing?additional_types=episode:
429: API rate limit exceeded

[2021-03-27 17:37:26,844] ERROR: Spotify could not find any track playing: Error in 
https ://api.spotify.com/v1/me/player/currently-playing?additional_types=episode:
429: API rate limit exceeded

[2021-03-27 17:37:27,356] ERROR: Spotify could not find any track playing: Error in
https ://api.spotify.com/v1/me/player/currently-playing?additional_types=episode:
429: API rate limit exceeded
```

**Reply by meltaxa** [reply]
I've made changes to the dev branch to throttle the API calls to Spotify. Let me know if this reduces or eliminates the API rate limit exceeded errors.

**Reply by r04drunn3r79** [reply]
You fixed the API rate limit issue. My musicfig.log is clean again. This version seems to work for me, I can select playlists with NFC tags, it plays the lightshow all the time and I have connected it to the bluetooth interface on my AVR. Thank you for your support.

Wicked app!

**Comment by miky2184** [comment]
Hi...thank u...i have only forked an existence project, i have added only the folder playlist feature...the dev branch is still in progress for the feature playlist folder

For the Spotify error...u have a premium or basic account in Spotify ??

Thank u
Bye bye



Inviato da iPhone

**Reply by r04drunn3r79** [reply]
I have a premium account.

**Comment by meltaxa** [comment]
Playlist feature was merged to master (see PR #13).

Usage example is in the tags.yml-sample file:

```
identifier:
    05631b62124:
        # LEGO minifigure. Play mp3 folder
        name: Peter Pan
        playlist: mp3/folder/name
        shuffle: on
```
